plenum/server/consensus/ordering_service.py

```diff
@@ -1,9 +1,991 @@
+from collections import defaultdict, OrderedDict
+from functools import partial
+from typing import Tuple, List, Set, Optional
+
+from sortedcontainers import SortedList
+
+from common.serializers.serialization import state_roots_serializer, invalid_index_serializer
+from crypto.bls.bls_bft_replica import BlsBftReplica
+from plenum.common.config_util import getConfig
+from plenum.common.constants import POOL_LEDGER_ID, SEQ_NO_DB_LABEL, ReplicaHooks, AUDIT_LEDGER_ID, TXN_TYPE
 from plenum.common.event_bus import InternalBus, ExternalBus
+from plenum.common.exceptions import SuspiciousNode, InvalidClientMessageException, SuspiciousPrePrepare, \
+    UnknownIdentifier
+# from plenum.common.messages.internal_messages import ViewChangeInProgress, LedgerSyncStatus, ParticipatingStatus
+from plenum.common.ledger import Ledger
+from plenum.common.messages.internal_messages import SuspiciousMessage, HookMessage, OutboxMessage
+from plenum.common.messages.node_messages import PrePrepare, Prepare, Commit, Reject
+from plenum.common.metrics_collector import MetricsName
+from plenum.common.request import Request
+from plenum.common.stashing_router import StashingRouter
+from plenum.common.timer import TimerService
+from plenum.common.txn_util import get_payload_digest
+from plenum.common.types import f
+from plenum.common.util import compare_3PC_keys, updateNamedTuple, SortedDict
+from plenum.server.batch_handlers.three_pc_batch import ThreePcBatch
 from plenum.server.consensus.consensus_data_provider import ConsensusDataProvider
+from plenum.server.propagator import Requests
+from plenum.server.replica import PP_APPLY_REJECT_WRONG, PP_APPLY_WRONG_DIGEST, PP_APPLY_WRONG_STATE, \
+    PP_APPLY_ROOT_HASH_MISMATCH, PP_APPLY_HOOK_ERROR, PP_SUB_SEQ_NO_WRONG, PP_NOT_FINAL, PP_APPLY_AUDIT_HASH_MISMATCH, \
+    PP_REQUEST_ALREADY_ORDERED, PP_CHECK_NOT_FROM_PRIMARY, PP_CHECK_TO_PRIMARY, PP_CHECK_DUPLICATE, \
+    PP_CHECK_INCORRECT_POOL_STATE_ROOT, PP_CHECK_OLD, PP_CHECK_REQUEST_NOT_FINALIZED, PP_CHECK_NOT_NEXT, \
+    PP_CHECK_WRONG_TIME, Replica, TPCStat, Stats
+from plenum.server.replica_validator_enums import INCORRECT_INSTANCE, DISCARD, INCORRECT_PP_SEQ_NO, ALREADY_ORDERED, \
+    STASH_VIEW, FUTURE_VIEW, OLD_VIEW, PROCESS, GREATER_PREP_CERT, STASH_CATCH_UP, CATCHING_UP, OUTSIDE_WATERMARKS, \
+    STASH_WATERMARKS
+from plenum.server.request_managers.write_request_manager import WriteRequestManager
+from plenum.server.suspicion_codes import Suspicions
+from stp_core.common.log import getlogger
+
+
+class ThreePCMsgValidator():
```

Comment:
Please make sure that `ThreePCMsgValidator` has the same tests as `ReplicaValidator` has (just copy-paste them and make necessary changes?)